File: scala/src/main/scala/com/digitalasset/fpbud17/Parser.scala
package com.digitalasset.fpbud17

import atto._, Atto._, compat.scalaz._

import scalaz.\/
import scalaz.syntax.applicative._

object Parser {
  import com.digitalasset.fpbud17.{ Scheme => S }

  def parseExpr(inp: String): String \/ S.Expr = expr.parseOnly(inp).either

  lazy val symbol = oneOf("!$%&|*+-/:<=?>@^_~#").named("symbol")
  lazy val spaces = skip(_.isWhitespace)

  lazy val parseString: Parser[S.Expr] = stringLiteral
    .named("string")
    .map(S.String)

  lazy val parseNumber: Parser[S.Expr] = bigInt
    .named("number")
    .map(S.Number)

  lazy val parseAtom: Parser[S.Expr] =
    ((letter | symbol) |@| stringOf(letterOrDigit | symbol))(_ +: _).map[S.Expr] {
        case "#t" => S.Bool(true)
        case "#f" => S.Bool(false)
        case a    => S.Atom(a)
    } named "atom"


  lazy val parseList: Parser[S.Expr] = sepBy(expr, whitespace).map[S.Expr](S.List).named("list")
  lazy val parseDottedPair: Parser[S.Expr] = {
    val p: Parser[S.Expr] = (token(expr) |@| (token(char('.')) ~> expr)) {
      case (head, tail) => S.Pair(head, tail)
    }
    p named "dotted-pair"
  }

  lazy val lists: Parser[S.Expr] = token(char('(')) ~> token(parseDottedPair | parseList) <~ char(')')

  lazy val expr: Parser[S.Expr] = (parseNumber | parseString | parseAtom | lists).named("expr")
}

File: haskell/src/Parser.hs
module Parser (
  parseExpr
  , readExpr
  , tryParseExpr
  ) where

import Text.ParserCombinators.Parsec 
import Scheme 

symbol :: Parser Char
symbol = oneOf "!#$%&|*+-/:<=>?@^_~"

parseString :: Parser Expr
parseString = do
                _ <- char '"'
                x <- many (noneOf "\"")
                _ <- char '"'
                return $ String x

parseNumber :: Parser Expr
parseNumber =  (Number . read) <$> many1 digit

  
parseAtom :: Parser Expr
parseAtom = do 
              first <- letter <|> symbol
              rest <- many (letter <|> digit <|> symbol)
              let atom = first:rest
              return $ case atom of 
                         "#t" -> Bool True
                         "#f" -> Bool False
                         _    -> Atom atom
 
parseList :: Parser Expr
parseList = do
     _ <- char '('
     es <- sepBy parseExpr spaces
     _ <- char ')'
     return $ List es

parsePair :: Parser Expr
parsePair = do
  _ <- char '('
  car <- parseExpr
  _ <- spaces
  _ <- char '.'
  _ <- spaces
  cdr <- parseExpr
  return $ Pair car cdr
  
parseExpr :: Parser Expr
parseExpr = parseAtom
         <|> parseString
         <|> parseNumber
         <|> (try parseList
              <|> parsePair)
        

readExpr :: String -> String
readExpr input = case parse (spaces >> parseExpr) "expression" input of
    Left err -> "No match: " ++ show err
    Right val -> "Found value: " ++ show val

tryParseExpr :: String -> Either String Expr
tryParseExpr input = case parse (spaces >> parseExpr) "expression" input of
  Left err -> Left $ "Failed to parse: " ++ (show err)
  Right e -> Right e
